GEODE-4233: Add an example for transactions.

PivotalSarge opened the pull request:
Add an example that uses five child processes to do one thousand concurrent increments on a counter.

upthewaterspout commented on transaction/README.md:
```diff
@@ -0,0 +1,60 @@
+configured across multiple servers and multiple clients may interact with that region independent of
+each other, data integrity relies on synchronization of modifications between all actors.
+
+An example of how data can become inconsistent during concurrent interaction is as follows:
```
Technically, this particular issue could be solved with a putIfAbsent since you are only modifying a single key, but I guess it's still a good example of transactions.

PivotalSarge replied:
I'm comfortable with a weaker case for transactions for a more concise example. Do you think this is a showstopper?

upthewaterspout replied:
It's not a showstopper, I get you are trying to make the example simple.

upthewaterspout commented on transaction/src/main/java/org/apache/geode_examples/transaction/Example.java:
```diff
@@ -0,0 +1,146 @@
+            cacheTransactionManager.begin();
+            final Integer oldValue = region.get(KEY);
+            final Integer newValue = oldValue + 1;
+            if (region.replace(KEY, oldValue, newValue)) {
```
This is weird that you are only committing if replace succeeds? If it fails you loop and call begin again without rolling back or committing? I'm not sure if the behavior of begin is defined in that case.

Using region.replace and a transaction here is basically wearing a belt and suspenders. Either one would get you the consistency you want here. Since you are demonstrating transactions, maybe just use a put?

upthewaterspout commented on transaction/src/main/java/org/apache/geode_examples/transaction/Example.java:
```diff
@@ -0,0 +1,146 @@
+            ++retries;
+          }
+        }
+      } catch (TransactionException te) {
```
Why catch these exceptions just to print a stack trace?

PivotalSarge commented:
@upthewaterspout I've fixed the logic around incrementing to use `Region.put` and rollback while I refactored the child process behavior into its own class.

upthewaterspout commented on transaction/src/main/java/org/apache/geode_examples/transaction/Incrementer.java:
```diff
@@ -0,0 +1,81 @@
+          incremented = true;
+        } catch (CommitConflictException cce) {
+          try {
+            cacheTransactionManager.rollback();
```
I don't think this rollback is necessary. If there is a CommitConflictException the transaction is  rolled back automatically I think.

PivotalSarge replied:
Roger that.

upthewaterspout commented on transaction/src/main/java/org/apache/geode_examples/transaction/Incrementer.java:
```diff
@@ -0,0 +1,81 @@
+          }
+        } catch (Exception e) {
+          try {
+            cacheTransactionManager.rollback();
```
Should we really catch unexpected exceptions here and try again? It might be better to just let the example fail in this case.

PivotalSarge replied:
Sure.

PivotalSarge commented:
I have uncovered some problems with forking child processes that need to be addressed.
